sale_coupon_multi_use/models/sale_order.py

```diff
@@ -1,32 +1,134 @@
         for order in self:
-            order = order.with_context(coupon_sale_order=order)
+            # Mimic same behavior as for single-user coupon.
+            order = order.with_context(**{ORDER_CTX_KEY: order})
             super(SaleOrder, order).action_confirm()
+            order._trigger_multi_use_coupons("used")
         return True
 
     def action_cancel(self):
         """Extend to pass coupon_sale_order context."""
-        # NOTE. Can't rely on coupon/SO relation, because
-        # applied_coupon_ids is o2m field, meaning same coupon used on
-        # another SO, would make lose relation on previous SO.
         for order in self:
-            order = order.with_context(coupon_sale_order=order)
+            order = order = order.with_context(**{ORDER_CTX_KEY: order})
             super(SaleOrder, order).action_cancel()
+            order._trigger_multi_use_coupons("new")
+        return True
+
+    def action_draft(self):
+        """Extend to pass coupon_sale_order context."""
+        super().action_draft()
+        for order in self:
+            # TODO: refactor this to use same approach as for
+            # action_confirm and action_cancel if Odoo fixes their bug.
+            # Now single use coupons are not reset back if you put SO to
+            # draft state.
+            coupons_multi_use = order.coupon_multi_use_ids
+            if coupons_multi_use:
+                coupons_multi_use.with_context(**{ORDER_CTX_KEY: order}).write(
+                    {"state": "used"}
+                )
         return True
+
+    def _get_coupons_from_2many_commands(self, commands):
+        ids = []
+        for cmd in commands:
+            if cmd[0] == 6:
+                ids.extend(cmd[2])
+            elif cmd[0] == 4:
+                ids.append(cmd[1])
+        return self.env["sale.coupon"].browse(ids)
+
+    def _get_multi_use_coupons_from_2many_commands(self, commands):
+        coupons = self._get_coupons_from_2many_commands(commands)
+        return coupons.filtered("multi_use")
+
+    def write(self, vals):
+        """Extend to add multi-use coupons."""
+        if vals.get("applied_coupon_ids"):
+            coupons_multi_use = self._get_multi_use_coupons_from_2many_commands(
+                vals["applied_coupon_ids"]
+            )
+            if coupons_multi_use:
+                vals["coupon_multi_use_ids"] = [(6, 0, coupons_multi_use.ids)]
+        return super().write(vals)
+
+
+class SaleOrderLine(models.Model):
+    """Extend to reverse relate with consumption line."""
+
+    _inherit = "sale.order.line"
+
+    def unlink(self):
+        """Extend to reactivate/reset removed multi-use coupons."""
+        related_program_lines = self.env["sale.order.line"]
+        # Reactivate coupons related to unlinked reward line
+        for line in self.filtered(lambda line: line.is_reward_line):
+            order = line.order_id
+            _filter = _get_order_line_coupons_filter(line)
```

Comment:
Remove this filter function and copy/paste the code in following filtered.
It's not easy to read like this, it will better to read/compare with your following code:
```
related_program = self.env["sale.coupon.program"].search(
    [("discount_line_product_id", "=", line.product_id.id)]
)
```
and
```
                related_program_lines |= (
                    order.order_line.filtered(
                        lambda l: l.product_id
                        == related_program.discount_line_product_id
                    )
                    - line
```

Reply:
Don't get it what you intend to do here. Your showed code does different thing than mine. Not related program lines.

Reply:
* Remove completely the function `_get_order_line_coupons_filter` (lines 6-7).
* Rewrite line 108:
`coupons_to_detach = order.coupon_multi_use_ids.filtered(lambda r: r.program_id.discount_line_product_id == line.product_id)`
* Rewrite lines 112-114:
`coupons_to_reactivate = (coupons_to_detach - order.applied_coupon_ids.filtered(lambda r: r.program_id.discount_line_product_id == line.product_id))`

Yes, you copy/paste two times the lambda filter, but the code will be more easy to read.